File: frontend-zk/src/api/zkAuth.ts
import { post } from "./client";

export interface ZKRegisterResponse {
  status: string;
}

export interface ZKChallengeResponse {
  nonce: string;
}

export interface ZKVerifyResponse {
  access_token: string;
  token_type: string;
}

export function register(
  publicKey: string,
  signature: string,
): Promise<ZKRegisterResponse> {
  return post<ZKRegisterResponse>("/zk-auth/register", {
    public_key: publicKey,
    signature,
  });
}

export function challenge(publicKey: string): Promise<ZKChallengeResponse> {
  return post<ZKChallengeResponse>("/zk-auth/challenge", {
    public_key: publicKey,
  });
}

export function verify(
  publicKey: string,
  nonce: string,
  signature: string,
): Promise<ZKVerifyResponse> {
  return post<ZKVerifyResponse>("/zk-auth/verify", {
    public_key: publicKey,
    nonce,
    signature,
  });
}

File: ndai/api/schemas/zk_auth.py
"""ZK authentication request/response schemas."""

from pydantic import BaseModel, Field


class ZKRegisterRequest(BaseModel):
    public_key: str = Field(..., min_length=64, max_length=64, pattern=r"^[0-9a-fA-F]{64}$")
    signature: str = Field(..., pattern=r"^[0-9a-fA-F]+$")
    alias: str | None = None


class ZKRegisterResponse(BaseModel):
    status: str


class ZKChallengeRequest(BaseModel):
    public_key: str = Field(..., min_length=64, max_length=64, pattern=r"^[0-9a-fA-F]{64}$")


class ZKChallengeResponse(BaseModel):
    nonce: str


class ZKVerifyRequest(BaseModel):
    public_key: str = Field(..., min_length=64, max_length=64, pattern=r"^[0-9a-fA-F]{64}$")
    nonce: str = Field(..., pattern=r"^[0-9a-fA-F]+$")
    signature: str = Field(..., pattern=r"^[0-9a-fA-F]+$")


class ZKVerifyResponse(BaseModel):
    access_token: str
    token_type: str = "bearer"
